### ISSUE TITLE
Flow Producer Design for cases where children jobs are dynamically defined by another worker

### ISSUE THREAD

**Issue by shaunakv1** [taskforcesh/bullmq#1408]
### The Problem

Please consider the following scenario.  

1. Say we have a system where  process starts with an input raster (image).    
2. We need to first preprocess this image that generates some metadata. This it self is a worker lets call it  `RasterWorker`
3. This metadata is used to generate many tiles from the raster in parallel using workers. Lets call this worker `TileWorker`
4.  Finally when all tiles are ready we need to  zip them up along with some more metadata. Lets call this `AssemblyWorker`

Now in this scenario , we need to run one `RasterWorker`,  use it's output to spawn multiple `TileWorker`s and  finally when all tiles are generated spawn the `AssemblyWorker`.

### My Current Solution

1. `RasterWorker` is it's own Queue and first the incoming image is processed via simply adding one job to this queue. 
2.  Then inside the `complete` event for `RasterWorker` we define a flow producer with `AssemblyJob` as parent and `TileJobs` as children

### Problems with this Solution

This works and my seem simple for this simplified scenario. However the code gets smelly quickly and jumps around.  Especially where FlowProducers are now spawned from another Jobs events.  This is true when you have more of these combinations in your desired flow. 

From what I understand this comes from the fact that   currently one parent in flow producer can have multiple children however, these children as a group can't have one (many) common child.

Ideally we would be able to define a flow that looks something like this: 

```
AssemblyJob (parent)
            |
            |
TileJobs (unknown number initially) (intermediate children and all parents to one child)
            |
            |
RasterJob (deepest child)
```

### Question

Am I stuck with my solution where writing different flow producers dynamically from other events? Or is there a better way to define this flow? 

One alternative I can think of is instead of having  multiple `TileWorkers` have a wrapper worker called `TilesPlanner`, which has `RasterWorker` as child and `AssemblyWorker` as a parent.  Then when the `RasterWorker` finishes and `TilesPlanner` starts processing, use it to spawn multiple `TileWorkers` and somehow wait for them to finish before marking `TilesPlanner` as done so that `AssemblyWorker` can kick in? 

If this is a better solution, what's a good design pattern to implement this? 
Also am I missing something here and can FlowProducer's current API already handle my use case that I am somehow missing?

**Comment by manast** [comment]
Have you read this blog post? https://blog.taskforce.sh/splitting-heavy-jobs-using-bullmq-flows/
The better solution is not using events for chaining jobs, instead add new jobs or flows from inside the process function of another job, by doing this you get a robust solution that will not have the edge case where the complete event may get lost or where adding a job from that event may fail and you get with a broken chain.

**Comment by rosslavery** [comment]
@manast Yes this is the approach we've adopted.

So in your example RasterWorker would create a flow with all of the children rather than using the complete event on the RasterWorker queue.

**Comment by shaunakv1** [comment]
@manast @rosslavery   Thanks for your inputs.  And thanks for pointing to the blog post. Its very useful and surprisingly similar to my example! 

It does make sense to have the new "nested" flow producer spawn from the worker itself, however we went with the events so that the actual spawning can take place from the main  express server environment. 

The main reason for us to not try and spawn the `TileJobs` from the first `RasterWorker` is  that Raster worker might be installed in a very thin (docker) environment that may not have all the dependencies that are needed to calculate data for the next steps! Think access to database,  binary dependencies etc. So our thought was to de-couple them using events. 

This is kind of why we were interested in trying to map the entire workflow in a FlowProducer.However if Events can be flaky, is there another "stable" way of de-coupling this without using events? 

Another issue with chaining future tasks from workers is that, it may not work very cleanly in more complex workflows we are trying to implement.  Above mentioned flow is just one of several other complex ones that are needed to process point clouds and LiDAR data.

**Comment by manast** [comment]
Instead of events you can use another intermediary Queue that is just used for notifications.
